File: backend/src/main/java/com/gxs/bank/model/SavingsAccount.java
package com.gxs.bank.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "savings_accounts")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class SavingsAccount {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @JsonIgnore
    private User user;

    @Column(name = "account_number", nullable = false, unique = true)
    private String accountNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "account_type", nullable = false)
    private AccountType accountType;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal balance;

    @Column(name = "interest_rate", nullable = false, precision = 5, scale = 2)
    private BigDecimal interestRate;

    @Column(name = "daily_interest_accrued", precision = 15, scale = 6)
    private BigDecimal dailyInterestAccrued;
    
    @Column(name = "ifsc_code", length = 11)
    private String ifscCode;

    @Column(name = "branch_name")
    private String branchName;

    @Column(name = "micr_code", length = 9)
    private String micrCode;

    @Column(name = "nominee_name")
    private String nomineeName;

    @Column(name = "daily_withdrawal_limit", precision = 15, scale = 2)
    private BigDecimal dailyWithdrawalLimit;

    @Column(name = "monthly_transfer_limit", precision = 15, scale = 2)
    private BigDecimal monthlyTransferLimit;

    @Column(name = "is_upi_enabled")
    private Boolean isUpiEnabled;

    @Column(name = "is_mobile_banking_enabled")
    private Boolean isMobileBankingEnabled;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Status status;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (balance == null) balance = BigDecimal.ZERO;
        if (dailyInterestAccrued == null) dailyInterestAccrued = BigDecimal.ZERO;
        if (status == null) status = Status.ACTIVE;
        if (accountType == null) accountType = AccountType.SAVINGS;
        if (ifscCode == null) ifscCode = "GXSB0000001";
        if (branchName == null) branchName = "GXS Digital Branch";
        if (dailyWithdrawalLimit == null) dailyWithdrawalLimit = new BigDecimal("200000.00");
        if (monthlyTransferLimit == null) monthlyTransferLimit = new BigDecimal("1000000.00");
        if (isUpiEnabled == null) isUpiEnabled = true;
        if (isMobileBankingEnabled == null) isMobileBankingEnabled = true;
    }

    public enum AccountType { SAVINGS, CURRENT, FIXED_DEPOSIT, RECURRING_DEPOSIT }
    public enum Status { ACTIVE, FROZEN, CLOSED }
}

File: backend-python/com/gxs/bank/model/SavingsAccount.py
from datetime import datetime
from decimal import Decimal
from enum import Enum
import uuid

from sqlalchemy import Boolean, Column, DateTime, Enum as SQLEnum, ForeignKey, Numeric, String
from sqlalchemy.orm import relationship

from com.gxs.bank.config.database import Base


class AccountType(str, Enum):
    SAVINGS = "SAVINGS"
    CURRENT = "CURRENT"
    FIXED_DEPOSIT = "FIXED_DEPOSIT"
    RECURRING_DEPOSIT = "RECURRING_DEPOSIT"


class Status(str, Enum):
    ACTIVE = "ACTIVE"
    FROZEN = "FROZEN"
    CLOSED = "CLOSED"


class SavingsAccount(Base):
    __tablename__ = "savings_accounts"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    userId = Column("user_id", String(36), ForeignKey("users.id"), nullable=False)
    user = relationship("User", lazy="select")

    accountNumber = Column("account_number", String(255), nullable=False, unique=True)
    accountType = Column("account_type", SQLEnum(AccountType), nullable=False, default=AccountType.SAVINGS)
    balance = Column(Numeric(15, 2), nullable=False, default=Decimal("0.00"))
    interestRate = Column("interest_rate", Numeric(5, 2), nullable=False)
    dailyInterestAccrued = Column("daily_interest_accrued", Numeric(15, 6), nullable=False, default=Decimal("0.000000"))
    ifscCode = Column("ifsc_code", String(11), nullable=False, default="GXSB0000001")
    branchName = Column("branch_name", String(255), nullable=False, default="GXS Digital Branch")
    micrCode = Column("micr_code", String(9), nullable=True)
    nomineeName = Column("nominee_name", String(255), nullable=True)
    dailyWithdrawalLimit = Column("daily_withdrawal_limit", Numeric(15, 2), nullable=False, default=Decimal("200000.00"))
    monthlyTransferLimit = Column("monthly_transfer_limit", Numeric(15, 2), nullable=False, default=Decimal("1000000.00"))
    isUpiEnabled = Column("is_upi_enabled", Boolean, nullable=False, default=True)
    isMobileBankingEnabled = Column("is_mobile_banking_enabled", Boolean, nullable=False, default=True)
    status = Column(SQLEnum(Status), nullable=False, default=Status.ACTIVE)
    createdAt = Column("created_at", DateTime, nullable=False, default=datetime.utcnow)


SavingsAccount.AccountType = AccountType
SavingsAccount.Status = Status
